rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,450 @@
+    }
+
+    pub async fn merge_with_options(
+        &self,
+        key: &[u8],
+        value: &[u8],
+        merge_opts: &MergeOptions,
+        write_opts: &WriteOptions
+    ) -> Result<(), SlateDBError> {
+        ...
+    }
+}
+
+impl WriteBatch {
+
+    pub fn merge(
+        &self,
+        key: &[u8],
+        value: &[u8]
+    ) -> Result<(), SlateDBError> {
+	    ...
+    }
+
+    pub fn merge_with_options(
+        &self,
+        key: &[u8],
+        value: &[u8],
+        merge_opts: &MergeOptions
+    ) -> Result<(), SlateDBError> {
+        ...
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct MergeOptions {
+    /// Merges with different TTLs will only be merged together during read but will be stored separately 
+    /// otherwise to allow them to expire independently. This ensures proper TTL handling while still
+    /// allowing merge operations to work correctly.
+    pub ttl: Ttl,
+}
+```
+
+### Merge Operator Interface
+
+We also need to introduce a new trait that will allow users to implement custom merging logic.
+
+```rust
+trait MergeOperator {
+
+    pub fn merge(
+        &self,
+        key: Bytes,
+        existing_value: Bytes,
+        value: Bytes
+    ) -> Result<Bytes, MergeOperatorError>;
+}
+```
+
+Initially, the implementation is limited to a single optional merge operator per database. The user must ensure that both the compactor and writer use the same merge operator to guarantee correct results.
+
+The merge operator can be implemented to route to different merge strategies based on key prefixes. Here's an example:
+
+```rust
+struct MyMergeOperator;
+
+impl MergeOperator for MyMergeOperator {
+
+    fn merge(
+        &self,
+        key: Bytes,
+        existing_value: Bytes,
+        value: Bytes
+    ) -> Result<Bytes, MergeOperatorError> {
+        if key.starts_with(b"list:") {
+            // concat values
+            Ok(...
+        } else if key.starts_with(b"min:") {
+            // calculate min
+            Ok(...)
+        } else {
+            Err(...)
+        }
+    }   
+}
+```
+
+It's **user's responsibility to ensure** that the merge operator is correctly implemented and **the same instance of the operator is used across all components** (compactor, writer, garbage collector).
+
+### Merge Operator Configuration
+
+User can provide their onw merge operator implementation via `DbOptions`. If not provided, the database will not support merge operations.
+
+```rust
+impl DbOptions {
+
+    #[serde(skip)]
+    /// The merge operator to use for the database. If not set, the database will not support merge operations.
+    pub merge_operator: Option<Arc<dyn MergeOperator + Send + Sync>>,
+}
+```
+
+### Extending TTL support
+
+The last public API change is extending the `Ttl` enum to support a new `ExpireAt(ts)` variant. This allows users to set a specific expiration time for a key, which overrides the default TTL behavior.
+
+```rust
+pub enum Ttl {
+    ...
+    ExpireAt(i64),
```

Comment:
One nit here (that I think might actually be important) is that I find having two variables named `ExpireAt` and `ExpireAfter` to be confusing; they're quite similarly named. Is there something better we can name these things to signal what they mean?

IIUC, ExpireAt lets use set a TTL once for a key and then it never changes (regardless of updates/writes) until the key has expired. ExpireAfter sets a TTL every time a put occurs, and only for the value/merge_value that is provided (so you could end up with multiple TTLs for a single key when writing merge values with different TTLs).

Some ChatGPT suggestions:

<img width="723" alt="Screenshot 2024-12-02 at 5 36 11 PM" src="https://github.com/user-attachments/assets/70957c15-c60e-489f-b3fb-9f2d60007765">

Some of these are wrong.. but I kinda like static/dynamic?

Reply:
Saw below:

```
2. **Per-Element TTL**
   - Each merge operation has its own independent TTL
   - Useful for expiring individual elements in collections
   - Supported via `Ttl::ExpireAfter(duration)` in `MergeOptions`

3. **Fixed Expiration Time**
   - Key expires at a specific time regardless of updates
   - Requires new `Ttl::ExpireAt(ts)` variant
   - All merges for a key must use the same expiration time
```

Perhaps something like "FixedTtl" and "PerElementTtl" or "PerPutTtl" or something like that?

Reply:
FYI motivation for using u64/i64 instead of `Duration` and `Instant` was (see https://github.com/slatedb/slatedb/pull/228#discussion_r1771959417) for (a) interoperability with other systems that choose different library to represent time and (b) for users of slateDB that choose to progress the clock with something other than time.

Not sure (b) is a particularly good motivation, and I'm not a rust expert enough to know whether (a) is legit, but ChatGPT made it seem like there were a few common libraries to represent time units - just wanted to chime in with context.

Reply:
> I did a little research and EXPIRE_AFTER + EXPIRE_AT doesn't seem to be completely new to the database community

It seems like in those systems `ExpireAfter` and `ExpireAt` behave the same way, it just allows you to specify the value in different ways (absolute vs. relative). That makes me think @criccomini's point that the naming is somewhat confusing because it differs from that.

It also makes me think that we should have different structures for Merges/Puts. I think both `ExpireAfter` and `ExpireAt` make a lot of sense to support for puts, with the traditional semantic of doing the same thing.

For merges, I think we should use a different structure -- something like:
```
MergeTtl {
  None,
  KeyExpiresAt(u64),
  MergeExpiresAfter(u64),
  MergeExpiresAt(i64) // maybe? this would be fixed expiration per-merge element and behaves the same as `MergeExpiresAfter` but with a timestamp specified instead of a duration
}
```

Reply:
> It seems like in those systems ExpireAfter and ExpireAt behave the same way, it just allows you to specify the value in different ways (absolute vs. relative). That makes me think @criccomini's point that the naming is somewhat confusing because it differs from that.

Yes! That's what got me confused. If `ExpireAt` is a point in time and `ExpireAfter` is a duration, and both only apply to the current `merge()` call, then that's intuitive to me, and I think the names are fine. This would correspond to `MergeExpiresAfter` and `MergeExpiresAt` in @agavra's example. But it sounds to me like @dmvk's version of `ExpiresAt` is actually what @agavra is calling `KeyExpiresAt`. 😵‍💫

Reply:
> It seems like in those systems ExpireAfter and ExpireAt behave the same way, it just allows you to specify the value in different ways (absolute vs. relative).

Yes! That's exactly what I'm trying to achieve here.

> This would correspond to MergeExpiresAfter and MergeExpiresAt in @agavra's example.

This is exactly what I'm trying to achieve. 

```
  /// indicates that the entire key (all merges) should expire at a given time
  /// SlateDB expects that all merges for the same key have the same 
  /// `KeyExpiresAt` value set
  KeyExpiresAt(u64),
```

I don't think we can and should support this, because it would put lot of complexity on compactions. This could be easily modeled in userspace using `MergeExpiresAt`.

---

This would actually leave us with the same datastructure (semantically) for puts and merges. We can of course split them into two so they can for example have more extensive documentation, but the logic is more or less the same.

WDYT?

Reply:
Ah ok, so I think perhaps what you're saying is that the only change is clarify that this
```
3. **Fixed Expiration Time**
   - Key expires at a specific time regardless of updates
   - Requires new `Ttl::ExpireAt(ts)` variant
   - All merges for a key must use the same expiration time
```
is implemented in userspace _using_ the new `Ttl:ExpireAt(ts)` syntax, it's not actually something we natively support in SlateDB.

Perhaps what that understand we're all saying the same thing 😇